{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a40a51bc-68e9-4d16-8550-9385805b33ed",
   "metadata": {},
   "source": [
    "# Collections & Data Structures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2690789-49ae-4e24-8457-ee5685c2a023",
   "metadata": {},
   "source": [
    "## Lists\n",
    "\n",
    "**Lists** are ordered, changeable collections. They are perfect for keeping track of sample names, concentrations, or a list of sequences.\n",
    "\n",
    "- **Creation**: Use square brackets [].\n",
    "- **Ordering**: Items stay in the order you put them in.\n",
    "- **Flexibility**: You can add, remove, or change items."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a801bf41-5879-4149-b1fc-7e5f45351160",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Sample List: ['Control_1', 'Control_2', 'Mutant_1_Changed', 'Mutant_2', 'Mutant_3']\n",
      "Total number of samples: 5\n"
     ]
    }
   ],
   "source": [
    "# Creating a list of samples\n",
    "samples = [\"Control_1\", \"Control_2\", \"Mutant_1\", \"Mutant_2\"]\n",
    "\n",
    "# Adding a new sample to the end\n",
    "samples.append(\"Mutant_3\")\n",
    "\n",
    "# Changing a specific entry (e.g., if you realized Mutant_1 was contaminated)\n",
    "samples[2] = \"Mutant_1_Changed\"\n",
    "\n",
    "print(f\"Current Sample List: {samples}\")\n",
    "print(f\"Total number of samples: {len(samples)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "695610b3-7971-4680-85bb-a3a78056fe10",
   "metadata": {},
   "source": [
    "### ðŸ§ª Useful List Methods\n",
    "\n",
    "| Method | Action |  Python Syntax Example |\n",
    "| :--- | :--- | :--- |\n",
    "| **`.append()`** | Adds an item to the end |  `samples.append(\"Mouse_5\")` |\n",
    "| **`.insert()`** | Adds item at specific index |  `samples.insert(0, \"Standard\")` |\n",
    "| **`.remove()`** | Removes a specific item |  `samples.remove(\"Degraded_DNA\")` |\n",
    "| **`.pop()`** | Removes item at index |  `next_up = samples.pop(0)` |\n",
    "| **`.sort()`** | Sorts the list in place |  `samples.sort()` |\n",
    "| **`.reverse()`**| Reverses the list order |  `samples.reverse()` |\n",
    "| **`.extend()`** | Merges two lists |  `plate1.extend(plate2)` |\n",
    "| **`.clear()`** | Removes all items |  `samples.clear()` |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fae2cfd-97f0-4f9c-bd3f-e8819f04593b",
   "metadata": {},
   "source": [
    "A very common mistake is trying to combine two lists using `.append()`.\n",
    "\n",
    "If you use `listA.append(listB)`, you get a list inside a list (a nested list).\n",
    "\n",
    "If you use `listA.extend(listB)`, you get one long, continuous list.\n",
    "\n",
    "`.extend()` is almost always what you want when merging data from different experimental runs!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "790fc5da-6b86-4faa-91d2-dc55fa31963d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final sample list: [45.2, 102.8, 63.7]\n"
     ]
    }
   ],
   "source": [
    "# An example of using the .append() method\n",
    "# Start with an empty list to store results\n",
    "sample = []\n",
    "\n",
    "sample.append(45.2)\n",
    "sample.append(102.8)\n",
    "sample.append(63.7)\n",
    "\n",
    "print(f\"Final sample list: {sample}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e25323b-cbae-4125-a2ee-eb103879f717",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Master Sample List: [45.2, 102.8, 63.7, 74.3, 115.9]\n"
     ]
    }
   ],
   "source": [
    "# An example of using the .extend() method\n",
    "samples = [45.2, 102.8, 63.7]\n",
    "\n",
    "# Samples from Day 2\n",
    "samples2 = [74.3, 115.9]\n",
    "\n",
    "# Use .extend() to merge the second list to the first one\n",
    "samples.extend(samples2)\n",
    "\n",
    "print(f\"Master Sample List: {samples}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40b68f17-75ff-4ca4-a43e-86fb0f651396",
   "metadata": {},
   "source": [
    "### âž• Merging with the Plus Operator\n",
    "\n",
    "While **`.extend()`** modifies the original list, the **`+`** operator creates a brand new list by joining two existing ones. This is particularly useful if you want to keep your original data separate and create a \"Master List\" for your analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8df18796-cb8f-40ab-992e-41dc8db22d9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original Controls: ['Control_A', 'Control_B']\n",
      "Original Treatments: ['Drug_10uM', 'Drug_50uM']\n",
      "Combined Dataset: ['Control_A', 'Control_B', 'Drug_10uM', 'Drug_50uM']\n"
     ]
    }
   ],
   "source": [
    "# Separate lists for different experimental groups\n",
    "control_group = [\"Control_A\", \"Control_B\"]\n",
    "treatment_group = [\"Drug_10uM\", \"Drug_50uM\"]\n",
    "\n",
    "# Create a NEW list by adding them together\n",
    "all_experimental_samples = control_group + treatment_group\n",
    "\n",
    "print(\"Original Controls:\", control_group)\n",
    "print(\"Original Treatments:\", treatment_group)\n",
    "print(\"Combined Dataset:\", all_experimental_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56a6c7c-e897-4dff-bf83-47a4a44d947c",
   "metadata": {},
   "source": [
    "Similar to Stirngs, the element of a list are accessed by an index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ff1fb51f-3aa0-4af3-89ab-e63eeec2c73f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The second element of a list: 63.7\n",
      "Next to the last element of a list: 102.8\n"
     ]
    }
   ],
   "source": [
    "print(f\"The second element of a list: {sample[1]}\")\n",
    "print(f\"Next to the last element of a list: {sample[-2]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a406c8-b507-40e5-9cbd-5256f01f76dd",
   "metadata": {},
   "source": [
    "You can delete elements of the list using the `del` command. **Note**: `del` command does **not** ise parentheses:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "774c4106-566c-4050-8c74-e1dca2b399c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[102.8, 63.7]\n"
     ]
    }
   ],
   "source": [
    "del sample[0]\n",
    "print(sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "964fbc37-c00c-4715-9d95-7bb75413fd71",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Dictionaries\n",
    "\n",
    "We often work with lists of numbers corresponding to specific genes, patients, or protein IDs. A **Dictionary** is the perfect data structure for this because it links a Key (the unique identifier) to a Value (the data).\n",
    "\n",
    "A **dictionary** is the ideal way to organize data from different brain regions, cell types, or experimental channels, where the \"Key\" is the specific identifier and the \"Value\" contains the measurement.\n",
    "\n",
    "- **Keys**: Must be unique and unchanging (Strings, Numbers, or Tuples).\n",
    "- **Values**: Can be anything (Numbers, Lists, even other Dictionaries).\n",
    "\n",
    "We might want to store:\n",
    "\n",
    "- words and their definitions\n",
    "- sample names and their coordinates\n",
    "- patient IDs and their health measurements (blood pressure, temperature, etc.)\n",
    "- protein sequence names and their sequences\n",
    "- DNA enzyme names and their motifs\n",
    "- codons and their associated amino acid residues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "974ede74-ad79-42f8-ac80-0cc120a57140",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A partial codon table\n",
    "genetic_code = {\n",
    "    \"AUG\": \"Methionine\",\n",
    "    \"CUU\": \"Leucine\",\n",
    "    \"UUU\": \"Phenylalanine\",\n",
    "    \"GGC\": \"Glycine\",\n",
    "    \"GCU\": \" Alanine\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aa74f7b-bef4-4a75-b702-b61f16336513",
   "metadata": {},
   "source": [
    "### ðŸ”‘ Dictionary Methods for Data Management\n",
    "\n",
    "| Method | Action |  Syntax Example |\n",
    "| :--- | :--- | :--- |\n",
    "| **`.keys()`** | Returns all Keys |  `my_dict.keys()` |\n",
    "| **`.values()`**| Returns all Values | `my_dict.values()` |\n",
    "| **`.items()`** | Returns Key-Value pairs | `my_dict.items()` |\n",
    "| **`.get()`** | Safe lookup |  `my_dict.get(\"Unit_99\", \"N/A\")` |\n",
    "| **`.update()`**| Merges dictionaries | `dict_1.update(dict_2)` |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da052a5a-fb75-4238-904f-f015a8dfb16c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes analyzed: ['GAPDH', 'TP53', 'BRCA1']\n"
     ]
    }
   ],
   "source": [
    "# --- Setup: Gene Expression dataset  ---\n",
    "gene_data = {\n",
    "    \"GAPDH\": 150.2,\n",
    "    \"TP53\": 12.8,\n",
    "    \"BRCA1\": 45.1\n",
    "}\n",
    "\n",
    "# .keys() - Retrieving Identifiers\n",
    "# Useful for: Getting a unique list of all genes detected in an assay.\n",
    "genes_detected = gene_data.keys()\n",
    "print(f\"Genes analyzed: {list(genes_detected)}\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a28fee84-b798-4148-80b4-43851f41431b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# .values() - Running Statistics\n",
    "# Useful for: Calculating global metrics like the mean expression across a sample.\n",
    "all_expression_values = gene_data.values()\n",
    "mean_expression = sum(all_expression_values) / len(all_expression_values)\n",
    "print(f\"Mean Expression: {mean_expression:.2f} FPKM\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7933ba49-0784-4ef4-b48d-9f9ecd9e02bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Search result for TP53: 12.8\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#  .get() - safe(r) aproach to get the value \n",
    "# Useful for: Searching for a gene in a large dataset where it might be missing.\n",
    "# Logic: .get(key, \"default_if_missing\")\n",
    "query_gene = \"TP53\"\n",
    "result = gene_data.get(query_gene, \"Gene not detected in this library\")\n",
    "print(f\"Search result for {query_gene}: {result}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b148f51e-f349-43fa-8f73-dca583b0a5ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final Combined Dataset: {'GAPDH': 150.2, 'TP53': 12.8, 'BRCA1': 45.1, 'MYC': 88.4, 'Sox2': 110.1}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# .update() - Merging Experimental Batches\n",
    "# Useful for: Combining results from two different sequencing runs.\n",
    "additional_gene_data = {\n",
    "    \"MYC\": 88.4,\n",
    "    \"Sox2\": 110.1\n",
    "}\n",
    "gene_data.update(additional_gene_data)\n",
    "\n",
    "print(f\"Final Combined Dataset: {gene_data}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00993345-3260-4402-92fa-5905ed08c72a",
   "metadata": {},
   "source": [
    "------\n",
    "\n",
    "## Tuples \n",
    "\n",
    "**A Tuple** is an ordered collection of items, just like a list, but with one major difference: it is immutable.\n",
    "\n",
    "*Syntax:* Defined using parentheses ( ) instead of square brackets.\n",
    "\n",
    "*Performance:* Slightly faster and more memory-efficient than lists.\n",
    "\n",
    "*Note:* Perfect for data that should be \"Read-Only\"\n",
    "\n",
    "\n",
    "### ðŸ”’ Tuple Methods Reference\n",
    "\n",
    "| Method | Action |Python Syntax Example |\n",
    "| :--- | :--- | :--- | \n",
    "| **`.count()`** | Counts occurrences  | `protocol.count(\"Control\")` |\n",
    "| **`.index()`** | Finds the first position | `markers.index(\"GFP\")` |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "acf41e0e-1998-4da3-aa3d-39d11701efb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of control replicates: 3\n"
     ]
    }
   ],
   "source": [
    "# A fixed experimental design: 3 Controls, 2 Treatments\n",
    "experimental_design = (\"Control\", \"Control\", \"Treatment_A\", \"Control\", \"Treatment_A\")\n",
    "\n",
    "# Count how many controls are in the design\n",
    "control_count = experimental_design.count(\"Control\")\n",
    "print(f\"Number of control replicates: {control_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3687a89f-f46d-4369-8f6f-c696e229d343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GFP is located at index: 1\n"
     ]
    }
   ],
   "source": [
    "# A list of markers used in an imaging panel\n",
    "markers = (\"DAPI\", \"GFP\", \"mCherry\", \"Cy5\")\n",
    "\n",
    "# Find where GFP is located in the sequence\n",
    "gfp_pos = markers.index(\"GFP\")\n",
    "print(f\"GFP is located at index: {gfp_pos}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d25d93d1-6d5d-486d-ba1d-a698313934d8",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### ðŸ“Š Summary of Python Collections\n",
    "\n",
    "| Structure | Syntax | Mutable? | Best Used For... |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **List** | `[ ]` | Yes | Ordered samples, list of sequences, ELISA readings. |\n",
    "| **Dictionary** | `{ : }`| Yes | Gene IDs to Expression values, Codon Tables, Metadata. |\n",
    "| **Tuple** | `( )` | No | Fixed coordinates, RGB color values, Constants. |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8816f294-57e3-4983-b3d0-b5db783f7c12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63c0637-7234-481b-bcec-ae70b8a44d71",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
